python-lib/pagination.py

```diff
@@ -95,31 +122,41 @@ def update_next_page(self, data, response_links=None):
         else:
             batch_size = 1
+        if self.skip_key:
+            self.skip = data.get(self.skip_key)
+            logger.info("update_next_page_per_page:skip=data[{}]={}".format(self.skip_key, self.skip))
+        self.records_to_skip = self.records_to_skip + batch_size
+        logger.info("update_next_page_per_page:records_to_skip={}, batch_size={}".format(self.records_to_skip, batch_size))
+
+
+    def update_next_page_link(self, data, response_links=None):
+        self.next_page_url = None
+        response_links = response_links or {}
+        next_link = response_links.get('next', {})
+        next_page_url = next_link.get("url")
+        self.is_first_batch = False
+        self.counter += 1
+        if next_page_url:
+            self.next_page_url = next_page_url
+            self.params_must_be_blanked = True
+        logger.info("update_next_page_link:next_link={}, next_page_url={}, params_must_be_blanked={}, next_page_number={}, counter={}".format(
+            next_link, self.next_page_url, self.params_must_be_blanked, self.next_page_number, self.counter
+        ))
+        self.data_is_list = False
         if self.next_page_key:
             next_page_path = extract_key_using_json_path(data, self.next_page_key)
             if self.next_page_url_base and next_page_path:
                 self.next_page_url = "/".join([self.next_page_url_base, next_page_path])
             else:
                 self.next_page_url = next_page_path
-            logger.info("update_next_page:next_page_url_base={}, next_page_path={}, next_page_url={}".format(
+            logger.info("update_next_page_link:next_page_url_base={}, next_page_path={}, next_page_url={}".format(
                 self.next_page_url_base, next_page_path, self.next_page_url
             ))
-        if self.skip_key:
-            self.skip = data.get(self.skip_key)
-            logger.info("update_next_page:skip=data[{}]={}".format(self.skip_key, self.skip))
-        if self.limit_key:
-            self.limit = data.get(self.limit_key)
-            logger.info("update_next_page:limit=data[{}]={}".format(self.limit_key, self.limit))
-        if self.total_key:
-            self.total = data.get(self.total_key)
-            logger.info("update_next_page:total=data[{}]={}".format(self.total_key, self.total))
-        self.records_to_skip = self.records_to_skip + batch_size
-        logger.info("update_next_page:records_to_skip={}, batch_size={}".format(self.records_to_skip, batch_size))
-        if self.total:
-            self.remaining_records = self.total - self.records_to_skip
-            logger.info("update_next_page:remaining_records={}, total={}, records_to_skip={}".format(
-                self.remaining_records, self.total, self.records_to_skip
-            ))
+
+    def update_next_page_default(self, data, response_links=None):
+        self.is_first_batch = False
+        self.counter += 1
+        self.next_page_number = self.next_page_number + 1
```

Comment:
The default is used in cases where no pagination mode is selected. I implemented the "no pagination mode" that way so that I don't have to switch cases in various place of the code according to whether or not the pagination is on enabled...

Reply:
Yep that makes sense - but I'm not sure why the method needs to do anything in that case - how the self.counter or self.next_page_number  get used - but I'm probably missing something....

Reply:
Regarding these 3 update next page methods, I started by duplicating them from the initial method, then remove things very conservatively, which might explain all the left overs... I cleaned most of it in a7b3509